### PR TITLE
[EXT2FS] Align with upstream and mark ReactOS diff

### PR DESCRIPTION
**Pull request by Vince1171**
## Purpose
add ifdefs to distinguish ReactOS code/fixes from the upstream code.
NOTE: there are no new features in this PR, we stay with upstream version 0.69.

JIRA issue: [CORE-18645](https://jira.reactos.org/browse/CORE-18645)

## Proposed changes

 * surround ReactOS modifications with ifdef
 * if upstream code was override, reintroduce it in #else section
 * re-align .rc version info with upstream (0.69)

### CODE REVIEW

File: drivers/filesystems/ext2/inc/linux/module.h [hunk 1]
```diff
@@ -79,6 +80,7 @@ RtlUlonglongByteSwap(
     IN ULONGLONG Source
 );
 #endif
```
Comment (JoachimHenze) [inline drivers/filesystems/ext2/inc/linux/module.h:82]:
```suggestion
#endif //#ifndef __REACTOS__
```

File: drivers/filesystems/ext2/src/devctl.c [hunk 2]
```diff
@@ -511,7 +527,11 @@ Ext2ProcessVolumeProperty(
             RtlCopyMemory(Property2->UUID, Vcb->SuperBlock->s_uuid, 16);
             Property2->DrvLetter = Vcb->DrvLetter;
 
+#ifdef __REACTOS__
             if (Property2->bHidingPrefix == Vcb->bHidingPrefix) {
+#else
+            if (Property2->bHidingPrefix = Vcb->bHidingPrefix) {
```
Comment (JoachimHenze) [inline drivers/filesystems/ext2/src/devctl.c:533]: Is that a bug in the original code, that we fixed in ReadctOS? Or is the original code correct here, and we just tried to mute some static code check and did it wrong?

File: drivers/filesystems/ext2/src/ext3/generic.c [hunk 3]
```diff
@@ -919,6 +924,77 @@ Ext2SaveBuffer( IN PEXT2_IRP_CONTEXT    IrpContext,
 }
 
 
+#else
```
Comment (JoachimHenze) [inline drivers/filesystems/ext2/src/ext3/generic.c:927]: Since the above is within #ifdef __REACTOS__ we could strip the 2empty lines 925 and 926.

File: drivers/filesystems/ext2/src/fsctl.c [hunk 4]
```diff
@@ -1799,7 +1799,7 @@ Ext2DeleteReparsePoint (IN PEXT2_IRP_CONTEXT IrpContext)
 
     BOOLEAN             FcbLockAcquired = FALSE;
     BOOLEAN             MainResourceAcquired = FALSE;
-
+
```
Comment (JoachimHenze) [inline drivers/filesystems/ext2/src/fsctl.c:1802]: Technically correct maybe, but I wouldn't do that. Open a PR in the source repo and remove it instead.

File: drivers/filesystems/ext2/src/devctl.c (hunk 2 above)
Comment (Vince1171) [reply drivers/filesystems/ext2/src/devctl.c:533]: it's a bug in the original code that we fixed indeed. After this PR, I'll try to upstream those fixes. but unfortunately, as stated in the Jira ticket, upstream seems to be abandoned. only https://github.com/bobranten/Ext4Fsd seems to still be around

File: drivers/filesystems/ext2/src/fsctl.c (hunk 4 above)
Comment (Vince1171) [reply drivers/filesystems/ext2/src/fsctl.c:1802]: Indeed

Comment (Vince1171) [reply drivers/filesystems/ext2/src/fsctl.c]: removed

File: drivers/filesystems/ext2/src/ext3/generic.c (hunk 3 above)
Comment (Vince1171) [reply drivers/filesystems/ext2/src/ext3/generic.c:925]: done :)

File: drivers/filesystems/ext2/inc/linux/module.h (hunk 1 above)
Comment (Vince1171) [reply drivers/filesystems/ext2/inc/linux/module.h:82]: done

File: drivers/filesystems/ext2/src/devctl.c (hunk 2 above)
Comment (HBelusca) [reply drivers/filesystems/ext2/src/devctl.c:533]: It would be worth looking at whether the ext4fsd of Bo Branten also covers ext2 and 3, and if so, how much does the code change wrt. that ext2/3 driver we have. In case of positive observations, maybe one could think about updating to that ext4fsd ?

Comment (Vince1171) [reply drivers/filesystems/ext2/src/devctl.c:533]: Yes, I'll try to have a look into it :)

Comment (gonzoMD) [reply drivers/filesystems/ext2/src/devctl.c:533]: > it's a bug in the original code that we fixed indeed. After this PR, I'll try to upstream those fixes. but unfortunately, as stated in the Jira ticket, upstream seems to be abandoned. only https://github.com/bobranten/Ext4Fsd seems to still be around

This also have this bug, so it's worth to send a PR. See https://github.com/bobranten/Ext4Fsd/blob/d9da413ad28afa3c0519d730da60f8598bbaea32/Ext4Fsd/devctl.c#L514

Comment (Vince1171) [reply drivers/filesystems/ext2/src/devctl.c:533]: after this PR is merged, I'll open a PR on ext4fsd with ReactOS fixes. Then I'll see to use Ext4fsd in ROS :)

File: drivers/filesystems/ext2/inc/linux/module.h [hunk 5]
```diff
@@ -59,7 +59,8 @@ extern "C" {
 #define RtlUlongByteSwap(_x)     _byteswap_ulong((_x))
 #define RtlUlonglongByteSwap(_x) _byteswap_uint64((_x))
 
-#elif !defined(__REACTOS__)
+#else
+#ifndef __REACTOS__
```
Comment (HBelusca) [inline drivers/filesystems/ext2/inc/linux/module.h:63]: why this change?

File: drivers/filesystems/ext2/inc/linux/module.h [hunk 6]
```diff
@@ -79,6 +80,7 @@ RtlUlonglongByteSwap(
     IN ULONGLONG Source
 );
 #endif
+#endif //#ifndef __REACTOS__
```
Comment (HBelusca) [inline drivers/filesystems/ext2/inc/linux/module.h:83]:
```suggestion
#endif //#ifndef __REACTOS__
#endif
```

File: drivers/filesystems/ext2/inc/linux/types.h [hunk 7]
```diff
@@ -27,14 +27,18 @@ typedef	signed   __int64	__s64;
 typedef	unsigned __int64	__u64;
 
 
+#ifdef __REACTOS__
```
Comment (HBelusca) [inline drivers/filesystems/ext2/inc/linux/types.h:30]: these typedefs weren't in the original code?

File: drivers/filesystems/ext2/src/ext4/ext4_extents.c [hunk 8]
```diff
@@ -14,9 +14,13 @@
  */
 
 #include "ext2fs.h"
+#ifdef __REACTOS__
 #include <linux/ext4.h>
```
Comment (HBelusca) [inline drivers/filesystems/ext2/src/ext4/ext4_extents.c:18]:
```suggestion
#include "linux/ext4.h"
```

File: drivers/filesystems/ext2/src/ext4/ext4_extents.c [hunk 9]
```diff
@@ -83,10 +87,16 @@ static inline void ext_debug(char *str, ...)
 {
 }
 #if TRUE
+#ifdef __REACTOS__
 #define EXT4_ERROR_INODE(inode, str, ...) do {                      \
             DbgPrint("inode[%p]: " str "\n", inode, ##__VA_ARGS__); \
         } while(0)
 #else
+#define EXT4_ERROR_INODE(inode, str, ...) do {                      \
```
Comment (HBelusca) [inline drivers/filesystems/ext2/src/ext4/ext4_extents.c:95]: Isn't that definition more correct than the ifdef'ed ReactOS one (minus the va_args that better need the ## in front of it)?
(or, is it possible to find why the ROS one removed the token pasting ## operator in front of str?)

File: drivers/filesystems/ext2/src/fileinfo.c [hunk 10]
```diff
@@ -10,8 +10,13 @@
 /* INCLUDES *****************************************************************/
 
 #include "ext2fs.h"
+#ifdef __REACTOS__
 #include <linux/ext4.h>
```
Comment (HBelusca) [inline drivers/filesystems/ext2/src/fileinfo.c:14]:
```suggestion
#include "linux/ext4.h"
```

File: drivers/filesystems/ext2/inc/linux/module.h (hunk 6 above)
Comment (Vince1171) [reply drivers/filesystems/ext2/inc/linux/module.h:83]: oops, indeed

File: drivers/filesystems/ext2/inc/linux/types.h (hunk 7 above)
Comment (Vince1171) [reply drivers/filesystems/ext2/inc/linux/types.h:30]: nope https://github.com/matt-wu/Ext3Fsd/blob/master/Ext3Fsd/include/linux/types.h#L31

File: drivers/filesystems/ext2/src/ext4/ext4_extents.c (hunk 8 above)
Comment (Vince1171) [reply drivers/filesystems/ext2/src/ext4/ext4_extents.c:18]: I didn't wanted to change anything in our code, except the ifdef.
but I suppose I can do that change

File: drivers/filesystems/ext2/inc/linux/module.h (hunk 5 above)
Comment (HBelusca) [reply drivers/filesystems/ext2/inc/linux/module.h:63]: ok keep it as you have done.

File: drivers/filesystems/ext2/src/ext4/ext4_extents.c (hunk 9 above)
Comment (Vince1171) [reply drivers/filesystems/ext2/src/ext4/ext4_extents.c:95]: I'm not familiar enough (yet) with this part of ext2fsd to answer this question.
all I can see is that it has been part of the integration of ext2fsd in our codebase https://github.com/reactos/reactos/commit/897634ce5bbb47b0f7c4ec48200baa77eb30b4f4 .
And I can trace back its origin to a patch in the related jira ticket https://jira.reactos.org/browse/CORE-10272 
https://jira.reactos.org/secure/attachment/31814/ext2fsd.patch  
but in the different conversation I haven't found an explanation about why they did that change

Comment (Vince1171) [reply drivers/filesystems/ext2/src/ext4/ext4_extents.c:95]: @HBelusca we cannot put ##str as str is always just a single string, one token, so no concatenation of token is possible here

Comment (HBelusca) [reply drivers/filesystems/ext2/src/ext4/ext4_extents.c:95]: yes actually you're correct. I've looked at the usages of `EXT4_ERROR_INODE` and it's used like:
```c
EXT4_ERROR_INODE(inode, "Some string with %s formatting", "something");
```
and so indeed, we cannot use the stringizer operator on `str`, otherwise it would re-stringize the string ^^

**Review by HBelusca — APPROVED** [review]

**Comment by HBelusca** [comment]
I suppose that those changes (restoring the proper ifdefs wrt. upstream) doesn't break reading EXT2/3 partitions with ReactOS.

**Comment by Vince1171** [comment]
> I suppose that those changes (restoring the proper ifdefs wrt. upstream) doesn't break reading EXT2/3 partitions with ReactOS.


no break indeed ^^

Screenshot from 2023-04-25 16-02-06

**Comment by HBelusca** [comment]
Thanks for the sync!
